### PR TITLE
Fixes required to build framework/library for mobile devices

### PR DESCRIPTION
**Pull request by drwasho**

### CODE REVIEW

File: mobile/node.go [hunk 1]
```diff
@@ -40,35 +53,40 @@ import (
...
-	"time"
 )
 
 type Node struct {
-	OpenBazaarNode *core.OpenBazaarNode
-	config         NodeConfig
-	cancel         context.CancelFunc
-	ipfsConfig     *ipfscore.BuildCfg
-	apiConfig      *schema.APIConfig
+	node       *core.OpenBazaarNode
```
[the remark this answers is not in the document]
Comment (drwasho) [reply mobile/node.go]: Renamed to `OpenBazaarNode`.

File: mobile/node.go [hunk 2]
```diff
@@ -40,35 +53,40 @@ import (
...
+	config     NodeConfig
+	cancel     context.CancelFunc
+	ipfsConfig *ipfscore.BuildCfg
+	apiConfig  *apiSchema.APIConfig
 }
 
-func NewNode(config NodeConfig) (*Node, error) {
+type Mobile struct{}
+
+func NewOB(repoPath string, authenticationToken string, testnet bool, userAgent string, walletTrustedPeer string, password string, mnemonic string) *Node {
```
[the remark this answers is not in the document]
Comment (drwasho) [reply mobile/node.go]: Done.

File: mobile/node.go [hunk 3]
```diff
@@ -40,35 +53,40 @@ import (
...
+	var m Mobile
+	node, err := m.NewNode(nodeconfig, password, mnemonic)
+	//n.Start()
+	if err != nil {
+		fmt.Println(err)
+	}
+	return node
+}
+
+func (m *Mobile) NewNode(config NodeConfig, password string, mnemonic string) (*Node, error) {
```
[the remark this answers is not in the document]
Comment (drwasho) [reply mobile/node.go]: Similar to the first comment, I believe we need to do it this way to overcome some of the limitations of gomobile.

From my limited understanding, I think it has something to do with this error if we use the current code:

```
gomobile bind -target=ios/arm64 github.com/OpenBazaar/openbazaar-go/mobile
gomobile: unsupported named type: struct{RepoPath string; AuthenticationToken string; Testnet bool; UserAgent string; WalletTrustedPeer string;DisableWallet bool; DisableExchangerates bool} / *types.Struct
```

Related:
- https://pliutau.com/gomobile-bind-types/
- https://godoc.org/golang.org/x/mobile/cmd/gobind#hdr-Type_restrictions

@hoffmabc did a lot of the work to get this initially up and running, based of Nick's help, so he may be able to provide additional insight

File: mobile/node.go [hunk 4]
```diff
@@ -78,52 +96,54 @@ func NewNode(config NodeConfig) (*Node, error) {
 	logger = logging.NewBackendFormatter(backendStdout, stdoutLogFormat)
 	logging.SetBackend(logger)
 
+	// Coin type
+	ct := wi.Bitcoin
+	/* 	switch walletCfg.Type {
```
[the remark this answers is not in the document]
Comment (drwasho) [reply mobile/node.go]: If removed, we get an error in `initializeRepo` as it requires the coin type as an argument. How should this be handled, or is it simpler to just leave this code in there?

File: mobile/node.go [hunk 5]
```diff
@@ -209,41 +230,45 @@ func NewNode(config NodeConfig) (*Node, error) {
 	}
 
 	// Wallet
-	var wallet wi.Wallet
-	var tp net.Addr
-	if config.WalletTrustedPeer != "" {
-		tp, err = net.ResolveTCPAddr("tcp", walletCfg.TrustedPeer)
+	/*
```
[the remark this answers is not in the document]
Comment (drwasho) [reply mobile/node.go]: Done.

File: mobile/node.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (drwasho) [reply mobile/node.go]: I've added some comments in the code to _hopefully_ help explain what's going.

File: vendor/github.com/OpenBazaar/multiwallet/client/mock.go [hunk 6]
```diff
@@ -208,7 +209,7 @@ var MockTransactions = []Transaction{
 			{
 				Txid:       "b466d034076ab53f4b019d573b6c68cf68c5b9a8cfbf07c8d46208d0fcf37762",
 				Vout:       0,
-				Sequence:   4294967294,
```
[the remark this answers is not in the document]
Comment (drwasho) [reply vendor/github.com/OpenBazaar/multiwallet/client/mock.go]: Done.

**Comment by drwasho** [comment]
@cpacia bumping this

**Comment by drwasho** [comment]
@cpacia

File: mobile/node.go [hunk 7]
```diff
@@ -5,13 +5,14 @@ import (
 	"os"
 	"path/filepath"
 
+	manet "gx/ipfs/QmRK2LxanhK2gZq6k6R7vk5ZoYZk8ULSSTB7FzDsMUX6CB/go-multiaddr-net"
+	ma "gx/ipfs/QmWWQ2Txc2c6tqjsBpzg5Ar652cHPGNsQQp2SejkNmkUMb/go-multiaddr"
+	ds "gx/ipfs/QmXRKBQA4wXP7xWbFiZsR1GP4HV6wMDQ1aWFxZZ4uBcPX9/go-datastore"
+
```
[the remark this answers is not in the document]
Comment (tyler-smith) [reply mobile/node.go]: go-imports defines a standard for organizing imports and we should use it.

File: mobile/node.go [hunk 8]
```diff
@@ -40,36 +53,45 @@ import (
...
+		RepoPath:            repoPath,
+		AuthenticationToken: "",
+		Testnet:             testnet,
+		UserAgent:           userAgent,
+		WalletTrustedPeer:   walletTrustedPeer,
+	}
+
+	// Use Mobile struct to carry config data
+	var m Mobile
+	node, err := m.NewNodeWithConfig(nodeconfig, password, mnemonic)
```
[the remark this answers is not in the document]
Comment (tyler-smith) [reply mobile/node.go]: How does `m` "carry config data"?

File: mobile/node.go [hunk 9]
```diff
@@ -40,36 +53,45 @@ import (
...
 	config         NodeConfig
 	cancel         context.CancelFunc
 	ipfsConfig     *ipfscore.BuildCfg
-	apiConfig      *schema.APIConfig
+	apiConfig      *apiSchema.APIConfig
 }
 
-func NewNode(config NodeConfig) (*Node, error) {
+// Mobile empty structure used to carry config data to route-around gomobile's type restrictions
+type Mobile struct{}
```
Comment (tyler-smith) [inline mobile/node.go]: What type restrictions? You mean for gomobile bind? How does this help with them?

**Comment by drwasho** [comment]
@cpacia The build was failing on the [latest commit](https://github.com/OpenBazaar/openbazaar-go/commits/multiwallet) of the multiwallet branch, which is what this PR is based off

File: mobile/node.go (hunk 9 above)
Comment (drwasho) [reply mobile/node.go]: This is what I posted to @cpacia in the first round of PR comments:

If we try and run gomobile without the changes in this PR, we get:

```
gomobile bind -target=ios/arm64 github.com/OpenBazaar/openbazaar-go/mobile
gomobile: unsupported named type: struct{RepoPath string; AuthenticationToken string; Testnet bool; UserAgent string; WalletTrustedPeer string;DisableWallet bool; DisableExchangerates bool} / *types.Struct
```

When we looked up this error, we found a few places talk about this issue:

https://pliutau.com/gomobile-bind-types/
https://godoc.org/golang.org/x/mobile/cmd/gobind#hdr-Type_restrictions

@hoffmabc did a lot of the work to get this initially up and running, based of Nick's help, so he may be able to provide additional insight

File: vendor/github.com/OpenBazaar/multiwallet/client/mock.go [hunk 10]
```diff
@@ -3,6 +3,7 @@ package client
 import (
 	"encoding/hex"
 	"errors"
+
```
[the remark this answers is not in the document]
Comment (drwasho) [reply vendor/github.com/OpenBazaar/multiwallet/client/mock.go]: Done.

File: mobile/node.go [hunk 11]
```diff
@@ -435,6 +422,7 @@ func (n *Node) Stop() error {
 	return nil
 }
 
+// initializeREpo create the database
```
[the remark this answers is not in the document]
Comment (drwasho) [reply mobile/node.go]: Done.

File: mobile/node.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (drwasho) [reply mobile/node.go]: @placer14 if we do this, will something need to change in line 93:

`func (m *Mobile) NewNodeWithConfig(config NodeConfig, password string, mnemonic string) (*Node, error) {...}`

File: mobile/node.go [hunk 12]
```diff
@@ -300,12 +284,13 @@ func NewNode(config NodeConfig) (*Node, error) {
 	}
 
 	if len(cfg.Addresses.Gateway) <= 0 {
-		return nil, errors.New("no gateway addresses configured")
+		return nil, errors.New("No gateway addresses configured")
 	}
 
-	return &Node{OpenBazaarNode: core.Node, config: config, ipfsConfig: ncfg, apiConfig: apiConfig}, nil
+	return &Node{config: config, OpenBazaarNode: core.Node, ipfsConfig: ncfg, apiConfig: apiConfig}, nil
```
[the remark this answers is not in the document]
Comment (drwasho) [reply mobile/node.go]: Can you please clarify @placer14?

File: mobile/node.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (placer14) [reply mobile/node.go]: I'm going to merge it as is for now. It's not ideal but there are better things to spend time on at the moment. I'll open an issue to track fixing this is we really want.

File: mobile/node.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (placer14) [reply mobile/node.go]: goimports doesn't seem to be working properly for me locally, but I'll manually arrange them here to match the standard it applies.

**Pull request merged**